book/src/proptest-derive/errors.md

```diff
@@ -0,0 +1,631 @@
+#[proptest(value = "MyStruct(42)")]
+struct MyStruct(u32);
+```
+
+These attributes are only applicable to fields.
+
+## E0008
+
+This error happens if `#[proptest(skip)]` is applied to an unskippable item.
+For example, struct fields cannot be skipped because Rust requires every field
+of a struct to have a value.
+
+Example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+struct WidgetContainer {
+    desired_widget_count: usize,
+    #[proptest(skip)]
+    widgets: Vec<Widget>,
+}
+```
+
+In general, the appropriate way to request proptest to not generate a field
+value is to use `#[proptest(value = "expr")]` to provide a fixed value
+yourself. For example, the above code could be properly written as follows:
+
+```rust
+#[derive(Debug, Arbitrary)]
+struct WidgetContainer {
+    desired_widget_count: usize,
+    #[proptest(value = "vec![]")] // Always generate an empty widget vec
+    widgets: Vec<Widget>,
+}
+```
+
+## E0009
+
+This error happens if `#[proptest(weight = <integer>)]` is applied to an item
+where this does not make sense, such as a struct field. For example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+struct Foo {
+    a: u32,
+    #[proptest(weight = 42)]
+    b: u32,
+}
+```
+
+The `weight` attribute only is sensible where proptest has a choice between
+multiple items, such as enum variants. In contrast, with struct fields proptest
+must provide a value for _every_ field so there is no "this-or-that" choice.
+
+## E0010
+
+This error occurs if `#[proptest(params = "type")]` and/or
+`#[proptest(no_params)]` are set on both an item and its parent.
+
+Example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+#[proptest(params = "String")]
+struct Foo {
+    #[proptest(no_params)]
+    bar: String,
+}
+```
+
+If the parent item has any explicit parameter configuration, it totally defines
+the parameters for the whole `Arbitrary` implementation and the child items
+must work with that and cannot specify their own parameters.
+
+## E0011
+
+This error occurs if `#[proptest(params = "type")]` is set on a field but no
+explicit strategy is configured with `#[proptest(strategy = "expr")]`. For
+example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+struct Foo {
+    #[proptest(param = "u8")]
+    some_string: String,
+}
+```
+
+This example illustrates why both must be specified: `String`'s arbitrary
+implementation takes a `proptest::string::StringParam`, but here we try to pass
+it a `u8`.
+
+While the generated code could work if the type given by `param` is the same as
+that for the default strategy, there would be no purpose in specifying the
+parameter type by hand; therefore specifying only `param` is in all cases
+forbidden.
+
+## E0012
+
+This error occurs if `#[proptest(filter = "expr")]` is set on an item, but the
+item containing it specifies a direct way to generate the whole value, which
+would thus occur without consulting the filter.
+
+Example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+enum Foo {
+    #[proptest(value = "Foo::V1(42)")]
```

Comment:
```suggestion
    #[proptest(value = "Foo::Bar(42)")]
```

It would be nice to have some "funner" names... `Bar` isn't that much better, but well, it fits :)

Reply:
I've applied this to most of the examples.

Reply:
Yeah; really like the improvements!